server/client.go

```diff
@@ -767,16 +767,23 @@ func (c *client) setPermissions(perms *Permissions) {
 
 // Check to see if we have an expiration for the user JWT via base claims.
 // FIXME(dlc) - Clear on connect with new JWT.
-func (c *client) checkExpiration(claims *jwt.ClaimsData) {
+func (c *client) setExpiration(claims *jwt.ClaimsData, validFor time.Duration) {
+	tn := time.Now().Unix()
+	expiresAt := time.Duration(0)
+	if claims.Expires > tn {
+		expiresAt = time.Duration(claims.Expires - tn)
+	}
 	if claims.Expires == 0 {
+		if validFor != 0 {
+			c.setExpirationTimer(validFor)
+		}
 		return
 	}
-	tn := time.Now().Unix()
-	if claims.Expires < tn {
-		return
+	if validFor != 0 && validFor < expiresAt {
```

Comment:
Humm... could you check something? It looks to me that `expiresAt` is considered a number of seconds, whereas `validFor` is a duration (expressed in nano seconds, result of end.Sub(start)). So I wonder in this case if you should not construct `expiresAt` directly as a time duration by `* time.Second` so the check is correct here and you don't need the `* time.Second` when passing to `setExpirationTimer()`. But maybe I am wrong and misunderstand that part.

Reply:
Moved `* time.Second` and added a unit test